File: ai_research/dataset_generation/infrastructure/storage.go
package infrastructure

import (
	"encoding/json"
	"github.com/kdunee/intentguard/dataset-generation/domain"
	"log"
	"os"
)

type OutputLine struct {
	Category string           `json:"category"`
	Examples []domain.Example `json:"examples"`
}

func WriteExamplesToOutput(category string, examples []domain.Example) {
	jsonBytes, err := json.Marshal(OutputLine{
		Category: category,
		Examples: examples,
	})
	if err != nil {
		log.Fatalf("Failed to marshal examples: %v", err)
	}

	f, err := os.OpenFile("output.jsonl", os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		log.Fatalf("Failed to open output file: %v", err)
	}
	defer f.Close()

	_, err = f.WriteString(string(jsonBytes) + "\n")
	if err != nil {
		log.Fatalf("Failed to write to output file: %v", err)
	}
}

File: ai_research/dataset_generation/infrastructure/storage.py
import json
from dataclasses import dataclass
from pathlib import Path

from ai_research.dataset_generation.domain.example import Example


@dataclass
class OutputLine:
    category: str
    examples: list[Example]

    def to_dict(self) -> dict[str, object]:
        return {
            "category": self.category,
            "examples": [example.to_dict() for example in self.examples],
        }


def write_examples_to_output(
    category: str,
    examples: list[Example],
    output_path: Path = Path("output.jsonl"),
) -> None:
    with output_path.open("a") as file:
        json.dump(OutputLine(category=category, examples=examples).to_dict(), file)
        file.write("\n")
